onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -528,74 +578,124 @@ void NchwcTransformerImpl::TransformBinary(Node& node, bool add_node) {
     nchwc_inputs.push_back(it->second.get());
   }
 
-  // Test if all of the NCHWc inputs have a compatible shape.
   auto* nchwc_input_0 = nchwc_inputs[0];
-  auto* nchwc_input_0_shape = input_defs[0]->Shape();
+  const int64_t channels = nchwc_inputs[0]->channels_;
+
+  // Test if all of the NCHWc inputs have an equal shape.
+  bool all_shapes_match = true;
+  auto* input_0_shape = input_defs[0]->Shape();
   for (size_t n = 1; n < input_defs_count; n++) {
     auto* nchwc_input_n = nchwc_inputs[n];
+    // Require that all inputs have the same logical number of channels.
+    if (nchwc_input_n->channels_ != channels) {
+      return;
+    }
     for (int i = 0; i < kNchwcDims; i++) {
       // Test if this dimension is derived from the same NodeArg.
       if (!nchwc_input_0->shape_.IsDimEqual(nchwc_input_n->shape_, i)) {
         // Check if ONNX shape inferencing has computed a precise dimension value.
-        auto* nchwc_input_n_shape = input_defs[n]->Shape();
-        if ((nchwc_input_0_shape == nullptr) || (nchwc_input_n_shape == nullptr)) {
-          return;
-        }
-        auto& nchwc_input_0_dim = nchwc_input_0_shape->dim(i);
-        auto& nchwc_input_n_dim = nchwc_input_n_shape->dim(i);
-        if (!utils::HasDimValue(nchwc_input_0_dim) ||
-            !utils::HasDimValue(nchwc_input_n_dim) ||
-            (nchwc_input_0_dim.dim_value() <= 0) ||
-            (nchwc_input_0_dim.dim_value() != nchwc_input_n_dim.dim_value())) {
-          return;
+        auto* input_n_shape = input_defs[n]->Shape();
+        if ((input_0_shape == nullptr) || (input_n_shape == nullptr)) {
+          all_shapes_match = false;
```

Comment:
>all_shapes_match = false; [](start = 10, length = 25)

nit:quit the loop earlier once all_shapes_match is false. maybe a helper function for the all shape match logic.

Reply:
Fair enough, i'll add that with the next NCHWc change. The outer loop still needs to continue though to check all inputs have the same channel count.